pubsub/psclient/client.go

```diff
@@ -1,97 +1,347 @@
+	}
+
+	go cl.receiver()
+
+	return cl
+}
+
+// Stop shutsdown the Client. If the websocket connection was created during
+// Client construction, it is is shutdown too.
+func (c *Client) Stop() {
+	log.Trace("Stopping psclient Client...")
+	// Close the user's receive channel.
+	close(c.recvMsgChan)
+
+	// Cancel the Context to stop the receiver loop.
+	c.shutdown()
+
+	// Close the websocket connection.
+	if c.ourConn {
+		if err := c.Conn.Close(); err != nil {
+			log.Errorf("Failed to Close websocket connection: %v", err)
+		}
 	}
 }
 
+// ClientMessage represents a message for a client connection.  The
+// (*Client).Receive method provides a channel such messages from the server.
+type ClientMessage struct {
+	EventId string
+	Message interface{}
+}
+
+// Receive gets a receive-only *ClientMessage channel, through which all
+// messages from the server to the client should be received.
+func (c *Client) Receive() <-chan *ClientMessage {
+	return c.recvMsgChan
+}
+
+// receiver receives and decodes messages from the server. It is to be run as a
+// goroutine by the constructor of the Client. The decoded ClientMessages are
+// sent on the recvMsgChan, the channel that is obtained via Receive.
+func (c *Client) receiver() {
+	defer c.shutdown()
+
+	for {
+		if c.ctx.Err() != nil {
+			log.Trace("receiver: context canceled...")
+			return
+		}
+
+		resp, err := c.receiveMsg()
+		if err != nil {
+			if pstypes.IsIOTimeoutErr(err) {
+				continue
+			}
+			log.Errorf("ReceiveMsg failed: %v", err)
+			return
+		}
+
+		msg, err := DecodeMsg(resp)
+		if err != nil {
+			log.Errorf("Failed to decode message: %v", err)
+			continue
+		}
+
+		switch m := msg.(type) {
+		case *pstypes.ResponseMessage:
+			log.Debugf("Response to %s request ID=%d received. Success = %v. Data: %v",
+				m.RequestEventId, m.RequestId, m.Success, m.Data)
+			respChan := c.responseChan(m.RequestId)
+			if respChan == nil {
+				log.Errorf("receiver failed to find request ID %d", m.RequestId)
+				continue
+			}
+			go func() {
+				respChan <- m // unbuffered
+				close(respChan)
+			}()
+			continue
+		case string:
+			// generic "message"
+			log.Debugf("Message (%s): %s", resp.EventId, m)
+		case int:
+			// e.g. "ping"
+			log.Debugf("Message (%s): %d", resp.EventId, m)
+		case *exptypes.WebsocketBlock:
+			log.Debugf("Message (%s): WebsocketBlock(hash=%s)", resp.EventId, m.Block.Hash)
+		case *exptypes.MempoolShort:
+			t := time.Unix(m.Time, 0)
+			log.Debugf("Message (%s): MempoolShort(numTx=%d, time=%v)",
+				resp.EventId, m.NumAll, t)
+		case *pstypes.TxList:
+			log.Debugf("Message (%s): TxList(len=%d)", resp.EventId, len(*m))
+		case *pstypes.AddressMessage:
+			log.Debugf("Message (%s): AddressMessage(address=%s, txHash=%s)",
+				resp.EventId, m.Address, m.TxHash)
+		default:
+			log.Debugf("Message of type %v unhandled.", resp.EventId)
+			continue
+		}
+
+		c.recvMsgChan <- &ClientMessage{
+			EventId: resp.EventId,
+			Message: msg,
+		}
```

Comment:
All of the broadcast messages are sent to the user via the `recvMsgChan`, obtained via `Receive`.